examples/task-allocator/task-allocator.c

```diff
@@ -0,0 +1,571 @@
+    InterlockedExchange(ptr, val);
+}
+
+static LONG atomic_load(atomic_int *ptr) {
+    return InterlockedCompareExchange(ptr, 0, 0);
+}
+
+static LONG atomic_fetch_add(atomic_int *ptr, LONG inc) {
+    return InterlockedExchangeAdd(ptr, inc);
+}
+
+static LONG atomic_fetch_sub(atomic_int *ptr, LONG dec) {
+    return atomic_fetch_add(ptr, -(dec));
+}
+
+static int pthread_create(pthread_t *out, void *unused,
+                          thread_ret_t (*func)(void *), void *arg) {
+    (void)unused;
+    HANDLE handle =
+        CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)func, arg, 0, NULL);
+    if (handle == NULL) {
+        return EAGAIN;
+    }
+
+    *out = handle;
+    return 0;
+}
+
+static int pthread_join(pthread_t thread, void *unused) {
+    (void)unused;
+    return (int)WaitForSingleObject(thread, INFINITE);
+}
+
+static int sched_yield(void) {
+    // https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-yieldprocessor
+    YieldProcessor();
+    return 0;
+}
+
+#else // ! _WIN32
+
+typedef void *thread_ret_t;
+
+#include <pthread.h>
+#include <stdatomic.h>
+
+#endif
+
+typedef pthread_t ggml_thread_t;
+
+//-----------------------------------------------------------------------------
+/// Most of the above codes are taken from
+/// https://github.com/ggerganov/llama.cpp/tree/master/ggml.c
+/// Copyright original authors.
+//-----------------------------------------------------------------------------
+
+#define MAX_THREADS 16
+
+struct task_allocator {
+    int nth;
+
+    int n_multiplier; // >= 1
+
+    atomic_int lock; // 0 unlocked, 1 locked
+
+    // total assigned.
+    atomic_int global_counter;
+
+    atomic_int thread_queue_heads[MAX_THREADS];
+    atomic_int thread_queue_tails[MAX_THREADS];
+};
+
+static void task_allocator_reset(struct task_allocator *a) {
+    for (int i = 0; i < a->nth; ++i) {
+        atomic_store(&a->thread_queue_heads[i], 0);
+        atomic_store(&a->thread_queue_tails[i], a->n_multiplier);
+    }
+
+    atomic_store(&a->lock, 0);
+    atomic_store(&a->global_counter, 0);
+}
+
+// NOTE: when nth == 1, n_multiplier is actually useless.
+static void task_allocator_init(struct task_allocator *a, int nth,
+                                int n_multiplier) {
+    GGML_ASSERT(nth > 0);
+    GGML_ASSERT(nth <= MAX_THREADS);
+    GGML_ASSERT(n_multiplier > 0);
+
+    a->nth = nth;
+    a->n_multiplier = nth == 1 ? 1 : n_multiplier;
+    task_allocator_reset(a);
+}
+
+// ith: worker id (start from 0).
+// chunk_idx and n_chunks will be updated.
+// chunk_idx is set as -1 when nothing to do.
+static void allocate_chunk(struct task_allocator *a, int ith, int *chunk_idx,
+                           int *n_chunks) {
+    GGML_ASSERT(ith >= 0 && ith < a->nth);
+
+    int M = a->n_multiplier;
+    int nth = a->nth;
+    int total_chunks = M * nth;
+
+    *chunk_idx = -1;
+    *n_chunks = total_chunks;
+
+    while (atomic_fetch_add(&a->lock, 1) != 0) { // lock
```

Comment:
This is a bad design, because:
1. individual thread MUST NOT be blocked on it's own queue.
2. should avoid spin lock in user space, think about a thread added the lock then be scheduled out. 
3. atomic_flag may be cheaper than the combination of atomic_fetch_add and atomic_fetch_sub.
     But, the contention is light so mutex is the best choice?